src/classes/CON_Preferred_Email_TEST.cls

```diff
@@ -0,0 +1,2215 @@
+        System.assertEquals(null, contactAfterBatchUpdateList[1].Email, 'Standard Email should be null');
+
+    }
+
+    /**************************************************************************************************************************
+    ******************************************* BULK FUNCTIONAL TESTS *********************************************************
+    **************************************************************************************************************************/
+
+    /*****************************************************************************************************************************
+    * @description Bulk test for Preferred Email in Insert context when Preferred Email Enforcement is enabled 
+    *****************************************************************************************************************************/
+    @isTest 
+    static void bulkTestToCoverAllPreferredEmailBeforeInsertScenariosWithEnforcementEnabled() {
+
+        UTIL_CustomSettingsFacade.getSettingsForTests(
+            new Hierarchy_Settings__c(Disable_Preferred_Email_Enforcement__c = false)
+        );
+
+        Contact testContact1 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact1.Email = '';
+        testContact1.Preferred_Email__c = 'Alternate Email';
+
+        Contact testContact2 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact2.Email = 'test@sf.org';
+
+        Contact testContact3 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact3.Email = '';
+        testContact3.AlternateEmail__c = 'test@sf.org';
+
+        Contact testContact4 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact4.Email = '';
+        testContact4.AlternateEmail__c = 'test@sf.org';
+        testContact4.WorkEmail__c = 'test@sf.org';
+
+        Contact testContact5 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact5.Email = '';
+        testContact5.Preferred_Email__c = 'Alternate Email';
+        testContact5.WorkEmail__c = 'test@sf.org';
+
+        Contact testContact6 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact6.Preferred_Email__c = 'Alternate Email';
+        testContact6.AlternateEmail__c = 'test@sf.org';
+        testContact6.Email = '';
+
+        Contact testContact7 = UTIL_UnitTestData_TEST.getUniqueContact(1);
+        testContact7.Preferred_Email__c = 'Test sf Email';
+        testContact7.WorkEmail__c = 'test@sf.org';
+
+        Contact testContact8 = testContact6.clone(false);
+        Contact testContact9 = testContact2.clone(false);
+
+        List<Contact> contactsToBeInserted = new List<Contact> {
+            testContact1, testContact2, testContact3, testContact4, testContact5, testContact6, testContact7, testContact8, testContact9
+                };
+    
+        List<Database.saveResult> results = Database.insert(contactsToBeInserted, false);
+
+        System.assertEquals(true, results[1].success);
```

Comment:
just making sure whether or not we need to worry about the order here? resulst[1] vs results[7] etc?

Reply:
Order is preserved here @lifewithryan i.e., elements in saveResult correspond to elements in contactsToBeInserted